Mathlib/Algebra/Lie/InvariantForm.lean

```diff
@@ -0,0 +1,207 @@
+  refine ⟨fun h x ↦ ?_, fun h x y z ↦ ?_⟩
+  · ext y z
+    rw [LieHom.lie_apply, LinearMap.sub_apply, Module.Dual.lie_apply, LinearMap.zero_apply,
+      LinearMap.zero_apply, h, sub_self]
+  · replace h := LinearMap.congr_fun₂ (h x) y z
+    simp only [LieHom.lie_apply, LinearMap.sub_apply, Module.Dual.lie_apply,
+      LinearMap.zero_apply, sub_eq_zero] at h
+    simp [← h]
+
+variable (hΦ_inv : Φ.lieInvariant L)
+
+/--
+The orthogonal complement of a Lie submodule `N` with respect to an invariant bilinear form `Φ` is
+the Lie submodule of elements `y` such that `Φ x y = 0` for all `x ∈ N`.
+-/
+@[simps!]
+def orthogonal (N : LieSubmodule R L M) : LieSubmodule R L M where
+  __ := Φ.orthogonal N
+  lie_mem {x y} := by
+    suffices (∀ n ∈ N, Φ n y = 0) → ∀ n ∈ N, Φ n ⁅x, y⁆ = 0 by
+      simpa only [LinearMap.BilinForm.isOrtho_def, -- and some default simp lemmas
+        AddSubsemigroup.mem_carrier, AddSubmonoid.mem_toSubsemigroup, Submodule.mem_toAddSubmonoid,
+        LinearMap.BilinForm.mem_orthogonal_iff, LieSubmodule.mem_coeSubmodule]
+    intro H a ha
+    rw [← neg_eq_zero, ← hΦ_inv]
+    exact H _ <| N.lie_mem ha
+
+@[simp]
+lemma orthogonal_toSubmodule (N : LieSubmodule R L M) :
+    (orthogonal Φ hΦ_inv N).toSubmodule = Φ.orthogonal N.toSubmodule := rfl
+
+lemma mem_orthogonal (N : LieSubmodule R L M) (y : M) :
+    y ∈ orthogonal Φ hΦ_inv N ↔ ∀ x ∈ N, Φ x y = 0 := by
+  simp [orthogonal, LinearMap.BilinForm.isOrtho_def, LinearMap.BilinForm.mem_orthogonal_iff]
+
+lemma orthogonal_disjoint
+    (Φ : LinearMap.BilinForm R L) (hΦ_nondeg : Φ.Nondegenerate) (hΦ_inv : Φ.lieInvariant L)
+    -- TODO: replace the following assumption by a typeclass assumption `[HasNonAbelianAtoms]`
+    (hL : ∀ I : LieIdeal R L, IsAtom I → ¬IsLieAbelian I)
+    (I : LieIdeal R L) (hI : IsAtom I) :
+    Disjoint I (orthogonal Φ hΦ_inv I) := by
+  rw [disjoint_iff, ← hI.lt_iff, lt_iff_le_and_ne]
+  suffices ¬I ≤ orthogonal Φ hΦ_inv I by simpa
+  intro contra
+  apply hI.1
+  rw [eq_bot_iff, ← lie_eq_self_of_isAtom_of_nonabelian I hI (hL I hI),
+      LieSubmodule.lieIdeal_oper_eq_span, LieSubmodule.lieSpan_le]
+  rintro _ ⟨x, y, rfl⟩
+  simp only [LieSubmodule.bot_coe, Set.mem_singleton_iff]
+  apply hΦ_nondeg
+  intro z
+  rw [hΦ_inv, neg_eq_zero]
+  have hyz : ⁅(x : L), z⁆ ∈ I := lie_mem_left _ _ _ _ _ x.2
+  exact contra hyz y y.2
+
+end ring
+
+section field
+
+variable {K L M : Type*}
+variable [Field K] [LieRing L] [LieAlgebra K L]
+variable [AddCommGroup M] [Module K M] [LieRingModule L M]
+
+variable [Module.Finite K L]
+variable (Φ : LinearMap.BilinForm K L) (hΦ_nondeg : Φ.Nondegenerate)
+variable (hΦ_inv : Φ.lieInvariant L) (hΦ_refl : Φ.IsRefl)
+-- TODO: replace the following assumption by a typeclass assumption `[HasNonAbelianAtoms]`
+variable (hL : ∀ I : LieIdeal K L, IsAtom I → ¬IsLieAbelian I)
+
+open FiniteDimensional Submodule in
+lemma orthogonal_isCompl_coe_submodule (I : LieIdeal K L) (hI : IsAtom I) :
```

Comment:
I don't think I'd bother with this lemma: it's the one below that we really want, right?

Hopefully using `LieSubmodule..isCompl_iff_coe_toSubmodule` it should be possible to survive with just `orthogonal_isCompl`?

Reply:
But this version is used multiple times below. So I would rather keep it.